packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -344,18 +344,24 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 	 */
 	protected abstract onDisconnect();
 
+	//* TODO: Add doc comments for rootMetadata throughout
+
 	/**
 	 * Submits a message by the local client to the runtime.
 	 * @param content - Content of the message
 	 * @param localOpMetadata - The local metadata associated with the message. This is kept locally by the runtime
 	 * and not sent to the server. This will be sent back when this message is received back from the server. This is
 	 * also sent if we are asked to resubmit the message.
 	 */
-	protected submitLocalMessage(content: any, localOpMetadata: unknown = undefined): void {
+	protected submitLocalMessage(
+		content: any,
+		localOpMetadata: unknown,
+		rootMetadata: unknown,
+	): void {
 		this.verifyNotClosed();
```

Comment:
I was wondering if we could just generate the `rootMetadata` here

Reply:
It's possible we could roundtrip decode/encode the content here and collect the outbound routes.  But that would be inefficient.  Maybe that's the default and then you can pass it in directly to avoid the perf hit.  I'll look into it.

Reply:
I'm curious as to how big of a performance issue this will become. What do you think? On encode, we would just store the information in some sort of data structure, and then once we submit, that's when we take whatever is in that data structure and package it with the op. Is that what you're concerned about?

Reply:
No that would be fine, that's what I'm imagining we'll do long-term.  Key point there is that it requires modifying all the `SharedObject` subclasses, which I think was your original point.

In the short term, since it's `content: any`, any generic measure is risky anyway (I didn't mention that in my previous comment).  That aside, the NAIVE generic approach would be to assume `content` is a POJO with serializable handles, then decode the handles, then encode them again and collect the list of handles.  That roundtrip is the perf concern.  But the real problem is the type safety - for example suppose a DDS applies some other transformation like compression on the op before calling `submitLocalMessage`.